### PR TITLE
Time dependent Hamiltonian | Part 1

### PR DESCRIPTION
**Pull request by abocquet**
Replaces https://github.com/pierreguilmin/torchqdynamics/pull/58. The philosophy is very close but the decorators logics is much lighter, as the only accepted variable is now time.

Only MERouchon1 takes advantage of the caching for now, as an example. Others will come in another PR, but are compatible with callable H

I'm not very satisfied with the way the duality Tensor/callable/list of callables is handled right now. Any idea is welcomed

For next PRs:

- [ ] piecewise constant H
- [ ] cache all numerical schemas

### CODE REVIEW

File: torchqdynamics/sesolve/propagator.py [hunk 1]
```diff
@@ -8,11 +8,9 @@ class SEPropagator(Solver):
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.H = self.H[:, None, ...]  # (b_H, 1, n, n)
-
     def run(self):
         y, t1 = self.y0, 0.0
         for t2 in tqdm(self.t_save, disable=not self.options.verbose):
-            y = torch.matrix_exp(-1j * self.H * (t2 - t1)) @ y
+            y = torch.matrix_exp(-1j * self.H(t2) * (t2 - t1)) @ y
```
Comment (gautierronan) [inline torchqdynamics/sesolve/propagator.py:14]: A priori the `propogator` method is always time independent. We will have to check this somewhere I guess, but it can be a separate PR.

**Review by gautierronan — APPROVED** [review]
I think we can merge this. And then we'll iterate in subsequent PRs.

File: tests/odeint/test_h_caching.py
Comment (pierreguilmin) [inline tests/odeint/test_h_caching.py:1]: This file should be in `tests/` directly.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Minor modifications, then OK to merge 🙌

**Review by gautierronan — APPROVED** [review]

**Review by pierreguilmin — APPROVED** [review]
Lezgooo 🥳
